File: cpp/day2/main.cpp
#include <charconv>
#include <fstream>
#include <vector>
#include <span>
#include <iostream>
#include <string>
#include <ranges>
#include <algorithm>

constexpr int sgn(int n) noexcept {
   if (n < 0)
      return -1;
   if (n == 0)
      return 1;
   return 0;
}

constexpr std::vector<int> calc_gradients(std::span<int> lineNums) {
   auto tmp = lineNums
      | std::views::slide(2)
      | std::views::transform([](auto rng) { auto iter = rng.begin(); return *iter++ - *iter; });

   std::vector<int> vecGradients{ tmp.begin(), tmp.end() };

   return vecGradients;
}

constexpr int part1(std::span<std::vector<int>> data) {
   auto ret = data
      | std::views::transform(&calc_gradients)
      | std::views::filter([](std::vector<int> lineGradients) {
         return std::ranges::all_of(lineGradients, [](int n) { return abs(n) >= 1 and abs(n) <= 3; });
      })
      | std::views::filter([](std::vector<int> lineGradients) {
         return std::ranges::all_of(lineGradients, [&lineGradients](int n) { return sgn(n) == sgn(lineGradients.front()); });
      });

   return std::ranges::distance(ret);
}

constexpr int part2(std::span<std::vector<int>> data) {
   auto ret = data
      | std::views::transform(&calc_gradients)
      | std::views::transform([](std::vector<int> vecGradients) {
         auto iter = std::ranges::find_if(vecGradients, [](int n) { return abs(n) < 1 or abs(n) > 3; });

         if (iter == vecGradients.end())
            return vecGradients;

         if (std::next(iter) != vecGradients.end()) {
            *iter = *iter + *std::next(iter);
            vecGradients.erase(std::next(iter));
         }

         return vecGradients;
      })
      | std::views::filter([](std::vector<int> vecGradients) {
         const auto numPos = std::ranges::distance(vecGradients | std::views::filter([](int n) { return n > 0; }));
         const auto numNeg = std::ranges::distance(vecGradients | std::views::filter([](int n) { return n < 0; }));
         return std::ranges::all_of(vecGradients, [](int n) { return abs(n) >= 1 and abs(n) <= 3; })
                  and (numPos < 2 or numNeg < 2);
      });

   return std::ranges::distance(ret);
}

int main() {
   std::ifstream f{ "input.txt" };
   std::string strBuffer{ std::istreambuf_iterator<char>{ f }, std::istreambuf_iterator<char>{} };
   f.close();

   auto tmp = strBuffer 
      | std::views::split('\n')
      | std::views::transform([](auto rng) { return std::string_view{ rng.begin(), rng.end() }; })
      | std::views::filter([](std::string_view line) { return not line.empty(); })
      | std::views::transform([](std::string_view strLine) { 
         return strLine 
            | std::views::split(' ') 
            | std::views::transform([](auto rng) {
               return std::string_view{ rng.begin(), rng.end() };
            });
      })
      | std::views::transform([](auto lineSplits) {
         std::vector<int> ret{};
         std::ranges::transform(lineSplits, std::back_inserter(ret), [](auto strNum) {
            int val{};
            std::from_chars(strNum.begin(), strNum.end(), val);
            return val;
         });
         return ret;
      });
   
   std::vector<std::vector<int>> data{ std::make_move_iterator(tmp.begin()), std::make_move_iterator(tmp.end()) };

   const auto nPart1 = part1(data);
   const auto nPart2 = part2(data);

   std::cout << "Result Part 1 = " << nPart1 << "\n";
   std::cout << "Result Part 2 = " << nPart2 << std::endl;
}

File: rust/day2/src/main.rs
use itertools::Itertools;

fn part1(data: &Vec<Vec<i32>>) -> usize {
    data
      .iter()
      .map(|row| row.iter().tuple_windows().map(|(x,y)| y-x).collect::<Vec<_>>())
      .filter(|row| row.iter().all(|x| (1..=3).contains(&x.abs())))
      .map(|row| {
          row
            .iter()
            .all(|x| x.signum() == row.first().unwrap().signum())
      })
      .filter(|v| *v)
      .count()
}

fn part2(data: &Vec<Vec<i32>>) -> usize {
    data
      .iter()
      .map(|row| row.iter().tuple_windows().map(|(x,y)| y-x).collect::<Vec<_>>())
      .map(|row| {
        let res = row.iter().find_position(|x| !(1..=3).contains(&x.abs()));
        match res {
            Some((idx, &num)) => {
                let mut new_row = row.clone();
                let new_val = if idx < row.len() - 1 {
                    num + row[idx+1]
                } else {
                    num
                };
                new_row[idx] = new_val;
                if idx < row.len() - 1 {
                    new_row.remove(idx+1);
                }
                new_row
            },
            None => row
        }
      })
      .filter(|row| {
          row.iter().all(|x| (1..=3).contains(&x.abs()))
          && 
          (row.iter().filter(|&&x| x < 0).count() < 2 || row.iter().filter(|&&x| x > 0).count() < 2)
       })
      .count()
}

fn main() {
//     let input = "7 6 4 2 1
// 1 2 7 8 9
// 9 7 6 2 1
// 1 3 2 4 5
// 8 6 4 4 1
// 1 3 6 7 9";
    let input = include_str!("../input.txt");

    let data = input
      .split("\n")
      .filter(|line| !line.is_empty())
      .map(|line| line.split(" "))
      .map(|str_nums| str_nums.map(|str_num| str_num.parse::<i32>().unwrap()).collect::<Vec<_>>())
      .collect::<Vec<_>>();   

    dbg!(&data); 

    let part1 = part1(&data);
    let part2 = part2(&data);

    println!("Result Part 1 = {}", part1);
    println!("Result Part 2 = {}", part2);
}
